auth/src/user.rs

```diff
@@ -94,20 +175,78 @@ pub struct User {
     pub name: AccountName,
     pub key: ApiKey,
     pub account_tier: AccountTier,
+    pub subscription_id: Option<SubscriptionId>,
 }
 
 impl User {
     pub fn is_admin(&self) -> bool {
         self.account_tier == AccountTier::Admin
     }
 
-    pub fn new(name: AccountName, key: ApiKey, account_tier: AccountTier) -> Self {
+    pub fn new(
+        name: AccountName,
+        key: ApiKey,
+        account_tier: AccountTier,
+        session_id: Option<SubscriptionId>,
+    ) -> Self {
         Self {
             name,
             key,
             account_tier,
+            subscription_id: session_id,
         }
     }
+
+    /// In case of an existing subscription, check if valid.
+    async fn subscription_is_valid(&self, client: &stripe::Client) -> Result<bool, Error> {
+        if let Some(subscription_id) = self.subscription_id.as_ref() {
+            let subscription = stripe::Subscription::retrieve(client, subscription_id, &[]).await?;
+            debug!("subscription: {:#?}", subscription);
+            return Ok(subscription.status == SubscriptionStatus::Active
+                || subscription.status == SubscriptionStatus::Trialing);
+        }
+
+        Ok(false)
+    }
+
+    // Synchronize the tiers with the subscription validity.
+    async fn sync_tier(&mut self, user_manager: &UserManager) -> Result<bool, Error> {
+        let subscription_is_valid = self
+            .subscription_is_valid(&user_manager.stripe_client)
+            .await?;
+        debug!(%subscription_is_valid, "to be or not to be");
+        if self.account_tier == AccountTier::Pro && !subscription_is_valid {
+            self.account_tier = AccountTier::PendingPaymentPro;
+            user_manager
+                .update_tier(&self.name, self.account_tier)
+                .await?;
+            return Ok(true);
+        }
+
+        if self.account_tier == AccountTier::PendingPaymentPro && subscription_is_valid {
+            self.account_tier = AccountTier::Pro;
+            user_manager
+                .update_tier(&self.name, self.account_tier)
+                .await?;
+            return Ok(true);
+        }
+
+        Ok(false)
+    }
+}
+
+impl FromRow<'_, SqliteRow> for User {
+    fn from_row(row: &SqliteRow) -> Result<Self, sqlx::Error> {
+        let subscription_id =
+            SubscriptionId::from_str(row.try_get("subscription_id").unwrap()).ok();
```

Comment:
Will this fail if the user doesn't have a subscription id?

Reply:
Good catch. We have a test that covers it: https://github.com/shuttle-hq/shuttle/blob/main/auth/tests/api/users.rs#L59. It looks like the `try_get(<column_name>)` will do as best as it can and will return an empty string after the `.unwrap()` in case the value is `NULL`, which would be an invalid subscription id and will cause a `None`, so that's why it wasn't an issue previously.

However, I will adjust this and avoid using the `unwrap()`.

Reply:
Adjusted this.